packages/ui/react-ui-editor/src/components/EditorToolbar/viewMode.ts

```diff
@@ -0,0 +1,48 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { type NodeArg } from '@dxos/app-graph';
+import { type ToolbarActionGroupProperties } from '@dxos/react-ui-menu';
+
+import { createEditorAction, createEditorActionGroup, type EditorToolbarState } from './util';
+import { translationKey } from '../../translations';
+
+const createViewModeGroupAction = (value: string) =>
+  createEditorActionGroup(
+    'viewMode',
+    {
+      variant: 'dropdownMenu',
+      applyActiveIcon: true,
+      selectCardinality: 'single',
+      value,
+    } as ToolbarActionGroupProperties,
+    'ph--eye--regular',
+  );
+
+const createViewModeActions = (value: string) =>
+  Object.entries({
+    preview: 'ph--eye--regular',
+    source: 'ph--pencil-simple--regular',
+    readonly: 'ph--pencil-slash--regular',
+  }).map(([viewMode, icon]) => {
+    return createEditorAction(
+      { type: 'view-mode', data: viewMode, checked: viewMode === value },
+      icon,
+      [`${viewMode} mode label`, { ns: translationKey }],
+      `view-mode--${viewMode}`,
+    );
+  });
+
+export const createViewMode = (state: EditorToolbarState) => {
+  const value = state.viewMode ?? 'source';
+  const viewModeGroupAction = createViewModeGroupAction(value);
+  const viewModeActions = createViewModeActions(value);
+  return {
+    nodes: [viewModeGroupAction as NodeArg<any>, ...viewModeActions],
+    edges: [
+      { source: 'root', target: 'viewMode' },
+      ...viewModeActions.map(({ id }) => ({ source: viewModeGroupAction.id, target: id })),
```

Comment:
I think this might be simplified by utilizing this convenience of the graph api https://github.com/dxos/dxos/blob/main/packages/sdk/app-graph/src/node.ts#L60

Reply:
Is there a good example of where this is used? It’s not clear from the type definition you linked which function or method does the work the comment describes.